File: electron/ipc/hooks.js
/**
 * Hooks IPC handlers — backend-driven (Hermes config.yaml).
 *
 * Helix's agent loop and tool execution run in the external `hermes`
 * subprocess, which ships its OWN hooks engine (`agent/shell_hooks.py`).
 * That engine registers hooks at GATEWAY STARTUP from the `hooks:` block of
 * Hermes' config.yaml, and (crucially) `pre_tool_call` hooks can BLOCK a tool
 * call — something an Electron-side observer can never do.
 *
 * So this module does NOT execute hooks. It is the persistence + restart
 * layer: it reads/writes the `hooks:` block (and `hooks_auto_accept`) in
 * `%LOCALAPPDATA%/hermes/config.yaml`, and asks the caller to restart the
 * gateway after a save so Hermes re-registers the hooks.
 *
 * Why `hooks_auto_accept: true`? Hermes only registers a hook if it is
 * allowlisted or `accept_hooks` is effective. In a non-TTY launch (Helix
 * spawns `hermes acp` without a terminal) there is no consent prompt, so the
 * hook is silently skipped unless `hooks_auto_accept: true` (or
 * `HERMES_ACCEPT_HOOKS=1` / `--accept-hooks`) is set. Since the user
 * explicitly configures these hooks in Settings, auto-accept is the correct
 * and necessary behavior.
 */
const { ipcMain } = require('electron')
const fs = require('fs')
const path = require('path')
const os = require('os')

// Event names MUST match Hermes' VALID_HOOKS (snake_case). Side-effect events
// only — transform_* / pre_gateway_dispatch are excluded because they expect a
// return value fed back into Hermes' control flow, which a plain command UI
// cannot satisfy.
const HOOK_EVENTS = [
  'pre_tool_call',
  'post_tool_call',
  'pre_verify',
  'on_session_start',
  'on_session_end',
  'on_session_finalize',
  'on_session_reset',
  'subagent_start',
  'subagent_stop',
  'pre_llm_call',
  'post_llm_call',
]

function _configYamlPath() {
  return path.join(os.homedir(), 'AppData', 'Local', 'hermes', 'config.yaml')
}

// ── YAML line editing (no js-yaml dep; preserves the rest of the file) ──────

// Remove a top-level `key:` block (and its indented children) from raw yaml.
function _stripTopLevel(yamlText, key) {
  const lines = yamlText.split('\n')
  const out = []
  let skip = false
  const re = new RegExp('^' + key + '(:\\s*|\\s)')
  for (const line of lines) {
    if (!skip && re.test(line)) { skip = true; continue }
    if (skip) {
      if (/^\S/.test(line)) { skip = false; out.push(line) } // next top-level key ends the block
      // else still inside the stripped block → drop
    } else {
      out.push(line)
    }
  }
  return out.join('\n')
}

// Quote a scalar the way Hermes/YAML expects; JSON.stringify yields a valid
// YAML double-quoted scalar whose JSON escaping round-trips via JSON.parse.
function _yamlString(v) {
  return JSON.stringify(String(v))
}

function _unyamlScalar(s) {
  if (s.length >= 2 && s[0] === '"' && s[s.length - 1] === '"') {
    try { return JSON.parse(s) } catch { return s.slice(1, -1) }
  }
  if (s.length >= 2 && s[0] === "'" && s[s.length - 1] === "'") {
    return s.slice(1, -1).replace(/''/g, "'")
  }
  return s
}

function _serializeHooks(hooksMap, autoAccept) {
  const lines = []
  lines.push(`hooks_auto_accept: ${autoAccept ? 'true' : 'false'}`)
  lines.push('hooks:')
  const events = Object.keys(hooksMap || {})
    .filter((e) => HOOK_EVENTS.includes(e) && Array.isArray(hooksMap[e]) && hooksMap[e].length)
  if (events.length === 0) {
    lines.push('  {}')
    return lines.join('\n') + '\n'
  }
  for (const ev of events) {
    lines.push(`  ${ev}:`)
    for (const h of hooksMap[ev]) {
      if (!h || typeof h.command !== 'string' || !h.command.trim()) continue
      lines.push(`    - command: ${_yamlString(h.command)}`)
      if (h.matcher && String(h.matcher).trim()) {
        lines.push(`      matcher: ${_yamlString(h.matcher)}`)
      }
      if (h.timeout != null && Number(h.timeout) > 0) {
        lines.push(`      timeout: ${Number(h.timeout)}`)
      }
    }
  }
  return lines.join('\n') + '\n'
}

function _parseHooks(text) {
  const lines = text.split('\n')
  let inHooks = false
  const hooks = {}
  let curEvent = null
  let curItem = null
  for (const raw of lines) {
    const line = raw.replace(/\r$/, '')
    if (!inHooks) {
      if (/^hooks:\s*\{\}\s*$/.test(line)) return {}
      if (/^hooks:\s*$/.test(line)) { inHooks = true; continue }
      continue
    }
    if (/^\S/.test(line)) break // next top-level key ends the hooks block
    const ev = line.match(/^  ([a-z_][a-z0-9_]*):\s*$/)
    if (ev) { curEvent = ev[1]; if (!hooks[curEvent]) hooks[curEvent] = []; curItem = null; continue }
    const item = line.match(/^    -\s*command:\s*(.*)$/)
    if (item && curEvent) {
      curItem = { command: _unyamlScalar(item[1].trim()), matcher: undefined, timeout: undefined }
      hooks[curEvent].push(curItem)
      continue
    }
    const mat = line.match(/^      matcher:\s*(.*)$/)
    if (mat && curItem) { curItem.matcher = _unyamlScalar(mat[1].trim()); continue }
    const to = line.match(/^      timeout:\s*(\d+)\s*$/)
    if (to && curItem) { curItem.timeout = Number(to[1]); continue }
  }
  return hooks
}

function loadHooks() {
  const yamlPath = _configYamlPath()
  let text = ''
  try { text = fs.readFileSync(yamlPath, 'utf-8') } catch {
    return { enabled: false, autoAccept: false, hooks: {} }
  }
  const hooks = _parseHooks(text)
  const enabled = Object.keys(hooks).length > 0
  let autoAccept = false
  const m = text.match(/^hooks_auto_accept:\s*(true|false)\s*$/m)
  if (m) autoAccept = m[1] === 'true'
  return { enabled, autoAccept, hooks }
}

function saveHooks(hooksMap, autoAccept) {
  const yamlPath = _configYamlPath()
  let text = ''
  try { text = fs.readFileSync(yamlPath, 'utf-8') } catch { text = '' }
  let stripped = _stripTopLevel(text, 'hooks')
  stripped = _stripTopLevel(stripped, 'hooks_auto_accept')
  const block = _serializeHooks(hooksMap, autoAccept)
  const merged = stripped.replace(/\s+$/, '') + '\n' + block
  const tmp = yamlPath + '.tmp'
  fs.writeFileSync(tmp, merged, 'utf-8')
  fs.renameSync(tmp, yamlPath)
  return { hooks: hooksMap, autoAccept }
}

// Restart callback is injected by main.js (it owns hermesProcess / startHermesGateway).
let _restartGateway = null

module.exports = {
  loadHooks,
  saveHooks,
  registerHooksHandlers(restartGateway) {
    if (typeof restartGateway === 'function') _restartGateway = restartGateway
    const channels = ['hooks:getConfig', 'hooks:setConfig']
    for (const ch of channels) {
      try { ipcMain.removeHandler(ch) } catch {}
    }
    ipcMain.handle('hooks:getConfig', async () => {
      try {
        return { ok: true, config: loadHooks() }
      } catch (e) {
        return { ok: false, error: String((e && e.message) || e) }
      }
    })
    ipcMain.handle('hooks:setConfig', async (_event, config) => {
      try {
        if (!config || typeof config !== 'object' || typeof config.hooks !== 'object') {
          return { ok: false, error: 'invalid hooks config' }
        }
        const enabled = config.enabled !== false
        const hooksMap = enabled ? (config.hooks || {}) : {}
        saveHooks(hooksMap, enabled)
        // Hermes registers hooks at GATEWAY STARTUP (register_from_config).
        // A restart is required for the new hooks to take effect.
        if (typeof _restartGateway === 'function') {
          try { await _restartGateway() } catch (e) {
            console.error('[hooks] gateway restart after save failed:', e.message)
          }
        }
        return { ok: true }
      } catch (e) {
        return { ok: false, error: String((e && e.message) || e) }
      }
    })
  },
}

File: src-tauri/src/hooks.rs
//! Hooks IPC — read/write the `hooks:` block + `hooks_auto_accept` in Hermes
//! config.yaml. Port of `electron/ipc/hooks.js`. Hermes registers hooks at
//! GATEWAY STARTUP, so a save restarts the gateway (acp mode only).
//!
//! NOTE: This file is byte-level YAML editing (no js-yaml dep) so the rest of
//! the config file is preserved untouched.

use crate::config::config_yaml_path;
use crate::gateway::{env_gateway_mode, kill_current, spawn_gateway};
use crate::state::AppState;
use serde_json::{json, Value};
use std::sync::Arc;
use tauri::State;

const HOOK_EVENTS: &[&str] = &[
    "pre_tool_call",
    "post_tool_call",
    "pre_verify",
    "on_session_start",
    "on_session_end",
    "on_session_finalize",
    "on_session_reset",
    "subagent_start",
    "subagent_stop",
    "pre_llm_call",
    "post_llm_call",
];

fn unyaml_scalar(s: &str) -> String {
    if s.len() >= 2 {
        if s.starts_with('"') && s.ends_with('"') {
            if let Ok(v) = serde_json::from_str::<String>(s) {
                return v;
            }
            return s[1..s.len() - 1].to_string();
        }
        if s.starts_with('\'') && s.ends_with('\'') {
            return s[1..s.len() - 1].replace("''", "'");
        }
    }
    s.to_string()
}

fn yaml_string(v: &str) -> String {
    serde_json::to_string(v).unwrap_or_else(|_| format!("\"{v}\""))
}

/// Remove a top-level `key:` block (and its indented children).
fn strip_top_level(yaml_text: &str, key: &str) -> String {
    let mut out: Vec<String> = Vec::new();
    let mut skip = false;
    for line in yaml_text.split('\n') {
        let at_top = line.starts_with(key)
            && line[key.len()..].starts_with(|c: char| c == ':' || c.is_whitespace());
        if !skip && at_top {
            skip = true;
            continue;
        }
        if skip {
            if line.starts_with(|c: char| !c.is_whitespace()) {
                skip = false;
                out.push(line.to_string());
            }
        } else {
            out.push(line.to_string());
        }
    }
    out.join("\n")
}

fn parse_hooks(text: &str) -> Value {
    let mut in_hooks = false;
    let mut hooks: Vec<Value> = Vec::new();
    let mut cur_event: Option<String> = None;
    let mut cur_item_idx: Option<usize> = None;
    for raw in text.split('\n') {
        let line = raw.trim_end_matches('\r');
        if !in_hooks {
            if line == "hooks: {}" {
                return json!({});
            }
            if line == "hooks:" {
                in_hooks = true;
                continue;
            }
            continue;
        }
        if line.starts_with(|c: char| !c.is_whitespace()) {
            break;
        }
        let trimmed = line.trim();
        // event: "  ev:" (exactly one trailing colon)
        if trimmed.ends_with(':') && !trimmed.starts_with('-') {
            let ev = trimmed[..trimmed.len() - 1].to_string();
            if !ev.is_empty()
                && ev.chars().all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '_')
            {
                cur_event = Some(ev);
                cur_item_idx = None;
                continue;
            }
        }
        // item: "    - command: ..."
        if let Some(rest) = trimmed.strip_prefix("- command:") {
            if let Some(ev) = &cur_event {
                let cmd = unyaml_scalar(rest.trim());
                let mut replaced = false;
                for e in hooks.iter_mut() {
                    if e.get("event").and_then(|v| v.as_str()) == Some(ev.as_str()) {
                        let arr = e.get_mut("items").and_then(|v| v.as_array_mut()).unwrap();
                        arr.push(json!({ "command": cmd }));
                        cur_item_idx = Some(arr.len() - 1);
                        replaced = true;
                        break;
                    }
                }
                if !replaced {
                    hooks.push(json!({ "event": ev, "items": [{ "command": cmd }] }));
                    cur_item_idx = Some(0);
                }
            }
            continue;
        }
        // matcher / timeout (attach to the last item of the current event)
        if let Some(rest) = trimmed.strip_prefix("matcher:") {
            if let (Some(ev), Some(idx)) = (&cur_event, cur_item_idx) {
                for e in hooks.iter_mut() {
                    if e.get("event").and_then(|v| v.as_str()) == Some(ev.as_str()) {
                        if let Some(arr) = e.get_mut("items").and_then(|v| v.as_array_mut()) {
                            if let Some(item) = arr.get_mut(idx) {
                                item["matcher"] = json!(unyaml_scalar(rest.trim()));
                            }
                        }
                        break;
                    }
                }
            }
            continue;
        }
        if let Some(rest) = trimmed.strip_prefix("timeout:") {
            if let (Some(ev), Some(idx)) = (&cur_event, cur_item_idx) {
                if let Ok(n) = rest.trim().parse::<u64>() {
                    for e in hooks.iter_mut() {
                        if e.get("event").and_then(|v| v.as_str()) == Some(ev.as_str()) {
                            if let Some(arr) = e.get_mut("items").and_then(|v| v.as_array_mut()) {
                                if let Some(item) = arr.get_mut(idx) {
                                    item["timeout"] = json!(n);
                                }
                            }
                            break;
                        }
                    }
                }
            }
            continue;
        }
    }
    let mut map = serde_json::Map::new();
    for e in hooks {
        let ev = e.get("event").and_then(|v| v.as_str()).unwrap_or("").to_string();
        let items = e.get("items").and_then(|v| v.as_array()).cloned().unwrap_or_default();
        map.insert(ev, json!(items));
    }
    json!(map)
}

#[tauri::command]
pub fn hooks_list() -> Value {
    let yaml_path = config_yaml_path();
    let text = match std::fs::read_to_string(&yaml_path) {
        Ok(t) => t,
        Err(_) => return json!({ "ok": true, "config": { "enabled": false, "autoAccept": false, "hooks": {} } }),
    };
    let hooks = parse_hooks(&text);
    let enabled = hooks.as_object().map(|m| !m.is_empty()).unwrap_or(false);
    let mut auto_accept = false;
    for line in text.split('\n') {
        if let Some(rest) = line.strip_prefix("hooks_auto_accept:") {
            auto_accept = rest.trim() == "true";
        }
    }
    json!({ "ok": true, "config": { "enabled": enabled, "autoAccept": auto_accept, "hooks": hooks } })
}

#[tauri::command]
pub fn hooks_save(state: State<'_, Arc<AppState>>, config: Value) -> Value {
    if !config.is_object() {
        return json!({ "ok": false, "error": "invalid hooks config" });
    }
    let hooks_v = config.get("hooks");
    if !hooks_v.map(|h| h.is_object()).unwrap_or(false) {
        return json!({ "ok": false, "error": "invalid hooks config" });
    }
    let enabled = config.get("enabled").and_then(|v| v.as_bool()).unwrap_or(true);
    let hooks_map = if enabled { hooks_v.unwrap().clone() } else { json!({}) };

    let yaml_path = config_yaml_path();
    let text = std::fs::read_to_string(&yaml_path).unwrap_or_default();
    let mut merged = strip_top_level(&text, "hooks");
    merged = strip_top_level(&merged, "hooks_auto_accept");
    let block = serialize_hooks(&hooks_map, enabled);
    let merged = merged.trim_end().to_string() + "\n" + &block;
    let tmp = yaml_path.with_extension("yaml.tmp");
    if std::fs::write(&tmp, &merged).is_err() {
        return json!({ "ok": false, "error": "write failed" });
    }
    if std::fs::rename(&tmp, &yaml_path).is_err() {
        return json!({ "ok": false, "error": "rename failed" });
    }
    // Hermes registers hooks at gateway startup — restart to apply (acp only).
    if env_gateway_mode() != "serve" {
        kill_current(&state);
        std::thread::sleep(std::time::Duration::from_millis(300));
        let _ = spawn_gateway(&state);
    }
    json!({ "ok": true })
}

fn serialize_hooks(hooks_map: &Value, auto_accept: bool) -> String {
    let mut lines = vec![format!("hooks_auto_accept: {}", if auto_accept { "true" } else { "false" })];
    lines.push("hooks:".to_string());
    if let Some(map) = hooks_map.as_object() {
        let mut wrote = false;
        for ev in HOOK_EVENTS {
            if let Some(arr) = map.get(*ev).and_then(|v| v.as_array()) {
                if arr.is_empty() {
                    continue;
                }
                lines.push(format!("  {ev}:"));
                for h in arr {
                    if !h.is_object() {
                        continue;
                    }
                    let cmd = h.get("command").and_then(|v| v.as_str()).unwrap_or("");
                    if cmd.trim().is_empty() {
                        continue;
                    }
                    lines.push(format!("    - command: {}", yaml_string(cmd)));
                    if let Some(m) = h.get("matcher").and_then(|v| v.as_str()) {
                        if !m.trim().is_empty() {
                            lines.push(format!("      matcher: {}", yaml_string(m)));
                        }
                    }
                    if let Some(t) = h.get("timeout").and_then(|v| v.as_u64()) {
                        if t > 0 {
                            lines.push(format!("      timeout: {t}"));
                        }
                    }
                }
                wrote = true;
            }
        }
        if !wrote {
            lines.push("  {}".to_string());
        }
    } else {
        lines.push("  {}".to_string());
    }
    lines.join("\n") + "\n"
}
